File: python/2512.py
from collections import defaultdict
from typing import List

class Solution:
    def topStudents(self, positive_feedback: List[str], negative_feedback: List[str], report: List[str], student_id: List[int], k: int) -> List[int]:
        score = defaultdict(int)
        for w in positive_feedback: score[w] = 3
        for w in negative_feedback: score[w] = -1
        students = [(-sum(score[w] for w in r.split()), i) for r, i in zip(report, student_id)]
        students.sort()
        return [i for _, i in students[:k]]

File: go/2512.go
package main

import (
	"sort"
	"strings"
)

func topStudents(positive_feedback []string, negative_feedback []string, report []string, student_id []int, k int) []int {
	score := map[string]int{}
	for _, w := range positive_feedback {
		score[w] = 3
	}
	for _, w := range negative_feedback {
		score[w] = -1
	}
	type pair struct{ score, id int }
	students := make([]pair, len(report))
	for i, r := range report {
		s := 0
		for _, w := range strings.Split(r, " ") {
			s += score[w]
		}
		students[i] = pair{s, student_id[i]}
	}
	sort.Slice(students, func(i, j int) bool {
		a, b := students[i], students[j]
		if a.score == b.score {
			return a.id < b.id
		}
		return a.score > b.score
	})
	ans := make([]int, k)
	for i, p := range students[:k] {
		ans[i] = p.id
	}
	return ans
}
